Added readme for Stacks-Subnets (#148)
hirosystems/stacks-subnets: README.md

# Subnets

> Subnets is a layer -2 scaling solution in the Stacks blockchain that offers low latency and high throughput workloads. It enables developers to build fast and reliable experiences on Stacks.

Reviewer: I think it should be

```
Subnets are a layer-2 scaling solution for the Stacks blockchain ...
```

## Overview

> The purpose of a Subnet is to be the optional “speed layer” for the Stacks blockchain. Subnets are designed to transact on Stacks assets, allowing users to move assets in and out of Subnets. While a user’s assets are in the Subnet, they trust the Subnet’s underlying consensus rules. The Subnet will then interact with the Stacks chain using a smart contract that is specific to that Subnet.

Reviewer: > "The purpose of a Subnet is to be the optional “speed layer” for the Stacks blockchain."

I think this sentence can just be dropped.

Also -- note for the rest of readme, I don't think subnets should be capitalized.

> The purpose of a Subnet is to be the optional “speed layer” for the Stacks blockchain. Subnets are designed to transact on Stacks assets, allowing users to move assets in and out of Subnets. While a user’s assets are in the Subnet, they trust the Subnet’s underlying consensus rules. The Subnet will then interact with the Stacks chain using a smart contract that is specific to that Subnet.

Reviewer: ```suggestion
Subnets are designed to transact on Stacks assets, meaning users can move assets in and out of subnets. While a user’s assets are in a subnet, they trust that subnet’s consensus rules. This subnet will interact with the Stacks chain using a smart contract specific to that subnet.
```

>[!Note]

> > The current implementation of Subnets supports a 2-phase commit amongst a fully-trusted pool of miners.

Reviewer: `supports` -> `uses`

Reviewer: `2-phase commit` -> `2-phase commit protocol`

Below are some of the features of Subnets:

> - Subnets have 4x times high throughput for transactions and may reduce confirmation time from 10 minutes to 1 minute.
> - Each Subnet is controlled by a single layer -1 smart contract. Interacting with a subnet is similar to interacting with stacks on a different network.
> - There can be many subnets in a single Stacks blockchain.
> - Each Subnet may use the same or different consensus rules.
> - Subnets can use larger blocks on the Stacks blockchain to increase the throughput of the transactions.
> - The current implementation of the Subnet consensus mechanism uses two-phase commit between a federated pool of miners.
> - To deposit into a Subnet, users submit an L1 transaction which involves the deposit method on a smart contract.
> - For withdrawal, after interacting with a Subnet, the user submits an L1 transaction which invokes the withdraw method on the same smart contract.
> - Subnets use BFT (Byzantine Fault Tolerance) protocol for mining. Each Subnet has a group of miners and each block of the Subnet is approved by a majority of the miners in that group. The miners on Subnet have complete control of the deposits and withdrawals arbitrarily.

Reviewer: ```suggestion
- Each subnet may define its own throughput settings. The default implementation should support at least 4x high throughput for transactions and may reduce confirmation time from 10 minutes to 1 minute.
- Interacting with a subnet is similar to interacting with a different Stacks network (e.g., testnet vs. mainnet).
- The Stacks blockchain can support many different subnets.
- Each Subnet may use the same or different consensus rules.
- This repository implements a consensus mechanism which uses two-phase commit among a federated pool of miners.
- To deposit into a subnet, users submit a layer-1 transaction to invoke the deposit method on that subnet's smart contract.
- For withdrawals, users commit the withdrawal on the subnet and then submit a layer-1 transaction to invoke the subnet's smart contract's withdraw method.
```

## Architecture

> Below is the architectural diagram that shows the interaction between Stacks and Subnet layers of a block chain.

Reviewer: ```suggestion
This diagram demonstrates the interaction between a subnet and the Stacks layer-1 chain.
```

![Architecture of Subnets.](docs/Images/subnet-architecture.png)

> When one miner proposes a block to the other miners, the other miners must approve the block before it can be committed to the Subnet.

Reviewer: ```suggestion
When a miner proposes a block to the other miners, the other miners must approve and sign the block before it can be committed to the subnet.
```

![Screenshot of Subnet miners proposing and approving the blocks.](docs/Images/subnet-miners-propose-and-approve.png)

### Trust models in Subnets

> Subnets use a federated trust system in its first version, and then iterate improvements on this system in subsequent versions.

Reviewer: ```suggestion
The current implementation of subnets uses a federated system of miners. This federation is fully-trusted, but future work on subnets will explore alternative trust models.
```

In a fully- trusted model:

> - Miners are responsible for issuing Subnet blocks.
> - Users can validate, but withdrawals are issued by a Subnet miner.
> - Trust can be federated with a BFT protocol of miners for block issuance.
> - This federation requires a majority of miners to approve withdrawals.

Reviewer: ```suggestion
- Miners are responsible for issuing subnet blocks.
- Users can validate, but subnet miners control withdrawals.
- Trust can be federated with a 2-phase commit and BFT protocol for miner block issuance.
- Federation requires a majority of miners to approve withdrawals.
```

In a fully trust-less model:

- Miners are responsible for issuing Subnet blocks.
- Users can **validate & issue withdrawals only if they correspond to a correct state** in the most recent valid subnet block.

> In many ways, the fully trust-less model is seen as the ideal Subnet model, since it is geared to a more decentralized approach while still having the benefits of Subnets. 

Reviewer: I think we can just drop the section on the fully trust-less model for now -- it isn't implemented yet in this repo.

>[!Note]

> > Participation in Trust model is optional, but when participating, the amount of trust can vary depending on the Subnet’s consensus model. 

Reviewer: Can drop this note.
